OpenRA.Game/Server/Server.cs

```diff
@@ -603,6 +605,35 @@ public void DispatchOrders(Connection conn, int frame, byte[] data)
 			else
 				DispatchOrdersToClients(conn, frame, data);
 
+			if (conn != null && data.Length > 0 && data[0] == (byte)OrderType.WinState)
+			{
+				// Skip order type (1B)
+				var ms = new MemoryStream(data, 1, data.Length - 1);
+				var reader = new BinaryReader(ms);
+				var pktPlayerWinstates = reader.ReadUInt64();
+
+				if (playerWinstates != pktPlayerWinstates)
+				{
+					var newLosses = playerWinstates ^ pktPlayerWinstates;
+					for (var i = 0; i < 64; i++)
+					{
+						if ((newLosses & (1UL << i)) != 0)
+						{
+							var lostClient = LobbyInfo.Clients.FirstOrDefault(c => c.Index == i);
+
+							// XXX: can the following happen?
+							if (lostClient == null)
+								continue;
+
+							var suffix = lostClient.Team != 0 ? " (Team {0})".F(lostClient.Team) : "";
+							Console.WriteLine("[{0}] {1}{2} is defeated", DateTime.Now.ToString(Settings.TimestampFormat), lostClient.Name, suffix);
```

Comment:
You need to check that all clients agree before trusting the input.

Reply:
This isn't really needed if we're just logging - a message "X is defeated" followed immediately by a desync (see my previous comment wrt guaranteeing this) is going to be pretty clear.

This *is* going to be important for replays or exposing a notification interface for external stat tracking etc, but this could also be solved by delaying a tick and making sure we haven't desynced.

Reply:
You can't rely on client side desyncing, a client can simply not send any sync packets and the desync will never be detected. So yes, you really have to compare the "winstate" before doing anything with it.

It isn't that important just for logging. But it wouldn't surprise me if somebody would start parsing the output and then feed it into a ladder system/whatever. Doing it correctly from the start would avoid that risk and also is also a prerequisite for any followup PR.

Reply:
So I did check if all clients agreed on the server side. I'm not sure if I did that correctly since I didn't use a dict like with the sync code to prevent an ever growing dict, only the last frame and state is kept.